From 96d15fb77c0cc8a39ecb7c4782e5f7669b568828 Mon Sep 17 00:00:00 2001
From: Vincent Bernat <vincent@bernat.im>
Date: Wed, 12 Apr 2017 11:24:07 +0200
Subject: [PATCH] Quick and dirty hack for JunOS compatibility

---
 bgpd/bgp_evpn.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/bgpd/bgp_evpn.c b/bgpd/bgp_evpn.c
index 20f54aaea6b5..5666a25db64e 100644
--- a/bgpd/bgp_evpn.c
+++ b/bgpd/bgp_evpn.c
@@ -292,7 +292,7 @@ form_auto_rt (struct bgp *bgp, struct bgpevpn *vpn,
   struct ecommunity_val eval;
   struct ecommunity *ecomadd;
 
-  encode_route_target_as ((bgp->as & 0xFFFF), vpn->vni, &eval);
+  encode_route_target_as ((bgp->as & 0xFFFF), vpn->vni | 0x10000000, &eval);
 
   ecomadd = ecommunity_new ();
   ecommunity_add_val (ecomadd, &eval);
-- 
2.11.0

